Source/WebKit/UIProcess/ios/WKContentView.mm

```diff
@@ -921,38 +930,45 @@ - (NSUInteger)_wk_pageCountForPrintFormatter:(_WKWebViewPrintFormatter *)printFo
             pageCount = _page->computePagesForPrintingiOS(frameID, printInfo);
     }
 
-    ASSERT(!_pdfPrintCompletionSemaphore);
-    if (isPrintingOnBackgroundThread)
-        _pdfPrintCompletionSemaphore = makeUnique<BinarySemaphore>();
-
-    // Begin generating the PDF in expectation of a (eventual) request for the drawn data.
-    _pdfPrintCallbackID = _page->drawToPDFiOS(frameID, printInfo, pageCount, [retainedSelf = retainPtr(self)](RefPtr<WebCore::SharedBuffer>&& pdfData) {
-        if (!pdfData || pdfData->isEmpty())
-            retainedSelf->_printedDocument = nullptr;
-        else {
-            auto data = pdfData->createCFData();
-            auto dataProvider = adoptCF(CGDataProviderCreateWithCFData(data.get()));
-            retainedSelf->_printedDocument = adoptCF(CGPDFDocumentCreateWithProvider(dataProvider.get()));
-        }
+    {
+        Locker locker { _pdfPrintCompletionLock };
 
-        if (auto pdfPrintCompletionSemaphore = std::exchange(retainedSelf->_pdfPrintCompletionSemaphore, nullptr))
-            pdfPrintCompletionSemaphore->signal();
+        ASSERT(!_pdfPrintCompletionCondition);
+        if (isPrintingOnBackgroundThread)
+            _pdfPrintCompletionCondition = makeUnique<Condition>();
+    }
 
-        retainedSelf->_pdfPrintCallbackID = 0;
+    ensureOnMainRunLoop([frameID, printInfo, pageCount, self, retainedSelf = retainPtr(self)] {
+        // Begin generating the PDF in expectation of a (eventual) request for the drawn data.
+        _pdfPrintCallbackID = _page->drawToPDFiOS(frameID, printInfo, pageCount, [retainedSelf](RefPtr<WebCore::SharedBuffer>&& pdfData) {
```

Comment:
What about the `_page->mainFrame()` above?  Should that also only be called from the main thread?

Reply:
Good catch – that will also need to be fixed.